### PR TITLE
Reject query if active worker nodes is insufficient

### PR DESCRIPTION
**Pull request by caithagoras**
Introduce config query.min-worker-nodes and reject queries if active
worker count is less than this value. However, once it is reached,
dropping below the threshold does not prevent query execution.

### CODE REVIEW

**Comment by highker** [comment]
Is that possible to queue the query instead of rejecting it? For large clusters, we expect to wait for a few seconds before a hundred workers can be online. To fail the queries may be too restrictive.

**Comment by electrum** [comment]
Don't end commit titles in periods. See https://chris.beams.io/posts/git-commit/

**Comment by electrum** [comment]
Please check Travis failures

File: presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.tests.tpch.TpchQueryRunner.createQueryRunner;
+import static io.airlift.testing.Assertions.assertContains;
+import static org.testng.Assert.fail;
+
+public class TestMinWorkerRequirement
+{
+
+    @Test
```
Comment (losipiuk) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java]: You can write this as 
```java
    @Test(expectedExceptions = PrestoException.class, expectedExceptionsMessageRegExp = "Insufficient active worker nodes")
    public void testInsufficientWorkerNodes()
             throws Exception
     {
         // TpchQueryRunner have 4 worker nodes by default
         try (DistributedQueryRunner queryRunner = createQueryRunner(ImmutableMap.of(), ImmutableMap.of("query.min-worker-nodes", "5"))) {
             queryRunner.execute("SELECT 1");
         }
     }
```

**Comment by caithagoras** [comment]
@highker This PR will mitigate the immediate problem, while we can look into ways to queue the request. And as you said, we expect most workers to be healthy in a few seconds, so not many queries will be impacted.

**Comment by raghavsethi** [comment]
@highker Let me know when it's ready for me to review.

File: presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+import com.google.common.collect.ImmutableMap;
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.tests.tpch.TpchQueryRunner.createQueryRunner;
+import static org.testng.Assert.fail;
+
+public class TestMinWorkerRequirement
+{
+
```
Comment (highker) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java]: remove empty lines

File: presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        // TpchQueryRunner have 4 worker nodes by default
+        try (DistributedQueryRunner queryRunner = createQueryRunner(ImmutableMap.of(), ImmutableMap.of("query.min-worker-nodes", "5"))) {
+            queryRunner.execute("SELECT 1");
+            fail("Expect to throw a PrestoException");
+        }
+    }
+
+    @Test
+    public void testSufficientWorkerNodes()
```
Comment (highker) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java]: Can we add a test with first having enough nodes and then some nodes crashed so that we are below the limit but can still run queries? One possible solution could be binding a mock `InternalNodeManager` with getNodes overridden by your own method. You are welcome to come up with easier solution.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 4]
```diff
@@ -120,10 +125,14 @@
 
     private final SessionPropertyManager sessionPropertyManager;
 
+    private final InternalNodeManager internalNodeManager;
+
     private final Map<Class<? extends Statement>, QueryExecutionFactory<?>> executionFactories;
 
     private final SqlQueryManagerStats stats = new SqlQueryManagerStats();
 
+    private final AtomicBoolean workersStarted = new AtomicBoolean(false);
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: The name is not quite accurate. I don't have a good suggestion for it. Maybe acceptQueries? @raghavsethi may have a better suggestion.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 5]
```diff
@@ -404,15 +426,15 @@ public QueryInfo createQuery(SessionSupplier sessionSupplier, String query)
         queryMonitor.queryCreatedEvent(queryInfo);
 
         queryExecution.addFinalQueryInfoListener(finalQueryInfo -> {
-                try {
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: looks like an unrelated change

**Comment by raghavsethi** [comment]
Travis is failing with multiple issues.

**Comment by caithagoras** [comment]
Fixed.

**Comment by raghavsethi** [comment]
@caithagoras Sorry I've been slammed recently. Will take a look at this Monday.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java [hunk 6]
```diff
@@ -283,4 +285,17 @@ public QueryManagerConfig setQueryExecutionPolicy(String queryExecutionPolicy)
...
         return this;
     }
+
+    @Min(1)
+    public int getMinWorkerNodes()
+    {
+        return minWorkerNodes;
+    }
+
+    @Config("query.min-worker-nodes")
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: This config name is pretty confusing - it seems like this is the min worker nodes required for every query (i.e. the min distribution required for the query). It also indicates that queries will not run unless this number is met (which is not true - this is only initial)

My idea would be to call this `query-manager.min-worker-nodes-initialized`.

Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: Add a one line `@ConfigDescription`.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 7]
```diff
@@ -342,6 +354,15 @@ public QueryInfo createQuery(SessionSupplier sessionSupplier, String query)
         QueryExecution queryExecution;
         Statement statement;
         try {
+            if (!acceptQueries.get()) {
+                if (internalNodeManager.getNodes(ACTIVE).size() < minWorkerNodes) {
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: I think this might include the coordinator?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java (hunk 6 above)
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: @martint @dain need your input on this.

File: presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+public class TestMinWorkerRequirement
+{
+    @Test(expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = "Insufficient active worker nodes")
+    public void testInsufficientWorkerNodes()
+            throws Exception
+    {
+        // TpchQueryRunner have 4 worker nodes by default
+        try (DistributedQueryRunner queryRunner = createQueryRunner(ImmutableMap.of(), ImmutableMap.of("query.min-worker-nodes", "5"))) {
+            queryRunner.execute("SELECT 1");
+            fail("Expect to throw a PrestoException");
```
Comment (raghavsethi) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java]: Be more specific with your error message

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java (hunk 7 above)
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: There could be multiple coordinators.

Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: I wonder what is the key here?

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 9]
```diff
@@ -120,10 +124,14 @@
 
     private final SessionPropertyManager sessionPropertyManager;
 
+    private final InternalNodeManager internalNodeManager;
+
     private final Map<Class<? extends Statement>, QueryExecutionFactory<?>> executionFactories;
 
     private final SqlQueryManagerStats stats = new SqlQueryManagerStats();
 
+    private final AtomicBoolean acceptQueries = new AtomicBoolean(false);
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: You can use the default/no-arg constructor since the default value is `false`.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 10]
```diff
@@ -342,6 +354,15 @@ public QueryInfo createQuery(SessionSupplier sessionSupplier, String query)
         QueryExecution queryExecution;
         Statement statement;
         try {
+            if (!acceptQueries.get()) {
+                if (internalNodeManager.getNodes(ACTIVE).size() < getMinWorkerNodesInitialized) {
+                    throw new PrestoException(SERVER_STARTING_UP, "Insufficient active worker nodes");
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: How about `The number of active worker nodes is insufficient`?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java [hunk 11]
```diff
@@ -54,6 +55,8 @@
     private Duration queryMaxRunTime = new Duration(100, TimeUnit.DAYS);
     private Duration queryMaxCpuTime = new Duration(1_000_000_000, TimeUnit.DAYS);
 
+    private int minWorkerNodesInitialized = 1;
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: We should find a better name for this config and related variables. How about `initialNumberOfWorkersRequired`?

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 12]
```diff
@@ -162,10 +171,13 @@ public SqlQueryManager(
 
         this.sessionPropertyManager = requireNonNull(sessionPropertyManager, "sessionPropertyManager is null");
 
+        this.internalNodeManager = requireNonNull(internalNodeManager, "internalNodeManager is null");
+
         this.minQueryExpireAge = config.getMinQueryExpireAge();
         this.maxQueryHistory = config.getMaxQueryHistory();
         this.clientTimeout = config.getClientTimeout();
         this.maxQueryLength = config.getMaxQueryLength();
+        this.getMinWorkerNodesInitialized = config.getMinWorkerNodesInitialized();
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: See my comment above about renaming this config and related variables.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 13]
```diff
@@ -342,6 +354,15 @@ public QueryInfo createQuery(SessionSupplier sessionSupplier, String query)
         QueryExecution queryExecution;
         Statement statement;
         try {
+            if (!acceptQueries.get()) {
+                if (internalNodeManager.getNodes(ACTIVE).size() < getMinWorkerNodesInitialized) {
+                    throw new PrestoException(SERVER_STARTING_UP, "Insufficient active worker nodes");
+                }
+                else {
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: don't need this else as previous if throws.

File: presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+public class TestMinWorkerRequirement
+{
+    @Test(expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = "Insufficient active worker nodes")
+    public void testInsufficientWorkerNodes()
+            throws Exception
+    {
+        // TpchQueryRunner have 4 worker nodes by default
+        try (DistributedQueryRunner queryRunner = createQueryRunner(ImmutableMap.of(), ImmutableMap.of("query.min-worker-nodes", "5"))) {
+            queryRunner.execute("SELECT 1");
+            fail("Expect to throw an Exception due to insufficient active worker nodes");
```
Comment (nezihyigitbasi) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java]: `Expected exception due to ...`

File: presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
+
+public class TestMinWorkerRequirement
+{
+    @Test(expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = "Insufficient active worker nodes")
+    public void testInsufficientWorkerNodes()
+            throws Exception
+    {
+        // TpchQueryRunner have 4 worker nodes by default
```
Comment (nezihyigitbasi) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java]: Instead of making this assumption why don't you update `createQueryRunner()` to get a number of workers parameter?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java (hunk 11 above)
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: Looking at the convention in QueryManagerConfig, we rarely have preposition in the config naming. What about query-manager.initial-min-workers-required?

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java (hunk 9 above)
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: I'm not sure if there is convention here but is this more readable I think?

**Comment by caithagoras** [comment]
Fixed. Thanks for reviewing.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java (hunk 5 above)
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: This was not correctly indentation.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java (hunk 9 above)
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: Most of the time we don't pass the default values to constructors as far as I can see. For this particular class if you do a quick search you will see that most usages follow this.

Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: There is a mixed use of the parameterized constructor in presto-main and presto-hive. I guess specifying false is to emphasize on the initial value. I won't mind removing it though.

**Comment by caithagoras** [comment]
@raghavsethi Done. PTAL.

**Comment by raghavsethi** [comment]
@caithagoras in general, please assign PRs back to me when you're done addressing comments. Makes it easier to keep track :)

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java [hunk 16]
```diff
@@ -283,4 +287,32 @@ public QueryManagerConfig setQueryExecutionPolicy(String queryExecutionPolicy)
...
     }
+
+    @Min(1)
+    public int getInitialMinWorkersRequired()
+    {
+        return initialMinWorkersRequired;
+    }
+
+    @Config("query-manager.initial-min-workers-required")
+    @ConfigDescription("minimum number of workers required before queries are accepted")
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: Minimum number of workers that must announce themselves before the cluster is ready to accept queries

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java [hunk 17]
```diff
@@ -283,4 +287,32 @@ public QueryManagerConfig setQueryExecutionPolicy(String queryExecutionPolicy)
...
+    }
+
+    @NotNull
+    public Duration getInitializationTimeout()
+    {
+        return initializationTimeout;
+    }
+
+    @Config("query-manager.initialization-timeout")
+    @ConfigDescription("Maximum amount of wait time after which queries will be accepted regardless of minimum worker requirement")
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: Maximum amount of time the cluster can remain in the `initializing` state (i.e. unable to accept queries)

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java (hunk 5 above)
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: It's great that you fixed this, but could you just do so in another commit?

File: presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java [hunk 18]
```diff
@@ -53,7 +53,7 @@
     public TestTpchDistributedStats()
             throws Exception
     {
-        runner = createQueryRunnerWithoutCatalogs(emptyMap(), emptyMap());
+        runner = createQueryRunnerWithoutCatalogs(emptyMap(), emptyMap(), 4);
```
Comment (raghavsethi) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java]: Can you just add another overloaded method that does not take a worker count (similar to what you did already)?

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 19]
```diff
@@ -342,6 +359,13 @@ public QueryInfo createQuery(SessionSupplier sessionSupplier, String query)
         QueryExecution queryExecution;
         Statement statement;
         try {
+            if (!acceptQueries.get()) {
+                if (nanosSince(initialNanos).compareTo(initializationTimeout) < 0 && internalNodeManager.getNodes(ACTIVE).size() < initialMinWorkersRequired) {
+                    throw new PrestoException(SERVER_STARTING_UP, "Insufficient active worker nodes");
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: "Cluster is still initializing, there are insufficient active worker nodes (%number) to run query"

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java [hunk 20]
```diff
@@ -283,4 +287,32 @@ public QueryManagerConfig setQueryExecutionPolicy(String queryExecutionPolicy)
...
         return this;
     }
+
+    @Min(1)
+    public int getInitialMinWorkersRequired()
+    {
+        return initialMinWorkersRequired;
+    }
+
+    @Config("query-manager.initial-min-workers-required")
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: Thought about this quite a bit and discussed @electrum. Let's call this initializationRequiredWorkers. Make the change to the property name also.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java [hunk 21]
```diff
@@ -283,4 +287,32 @@ public QueryManagerConfig setQueryExecutionPolicy(String queryExecutionPolicy)
...
+    @ConfigDescription("Minimum number of workers that must announce themselves before the cluster is ready to accept queries")
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: "Minimum number of workers that must be available before the cluster will accept queries"

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java [hunk 22]
```diff
@@ -283,4 +287,32 @@ public QueryManagerConfig setQueryExecutionPolicy(String queryExecutionPolicy)
...
+    }
+
+    @NotNull
+    public Duration getInitializationTimeout()
+    {
+        return initializationTimeout;
+    }
+
+    @Config("query-manager.initialization-timeout")
+    @ConfigDescription("Maximum amount of time the cluster can remain in the initializing state (i.e. unable to accept queries)")
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: "After this time, the cluster will accept queries even if the minimum required workers are not available"

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 23]
```diff
@@ -342,6 +359,16 @@ public QueryInfo createQuery(SessionSupplier sessionSupplier, String query)
         QueryExecution queryExecution;
         Statement statement;
         try {
+            if (!acceptQueries.get()) {
+                int activeWorkerCount = internalNodeManager.getNodes(ACTIVE).size();
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: Did you check that this does not include coordinator?

Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: I guess we should also handle the case for `node-scheduler.include-coordinator=true` similar to how we handle it [here](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java#L57-L60).

Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: Done

File: presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+import com.google.common.collect.ImmutableMap;
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.tests.tpch.TpchQueryRunner.createQueryRunner;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
+
+public class TestMinWorkerRequirement
+{
+    @Test(
```
Comment (raghavsethi) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestMinWorkerRequirement.java]: Don't wrap these at all - it's OK to exceed line length.

**Comment by caithagoras** [comment]
This one Travis build keeps failing but when I run locally with "mvn install -T1C -e -pl presto-tests", all tests passed. Is this a known issue or am I missing something here?
